unit_tests/src/test_json_apis_from_client.cpp

```diff
@@ -0,0 +1,110 @@
+
+#include <boost/filesystem.hpp>
+#include <fmt/format.h>
+
+#include <chrono>
+#include <string_view>
+#include <string>
+#include <stdexcept>
+
+#include <unistd.h>
+#include <cstdlib>
+
+#include "json.hpp"
+
+
+namespace fs = irods::experimental::filesystem;
+
+using nlohmann::json;
+using namespace std::string_literals;
+
+namespace {
+
+    class error_setting_environment : public std::runtime_error {
+        public:
+        explicit error_setting_environment(const std::string & msg) : std::runtime_error{msg} {
+        }
+    };
+
+    void change_environment_variable (const char* env_var_name, char* & charptr_tmp, const char* new_value) {
+        charptr_tmp = getenv(env_var_name);
+        if (charptr_tmp) { charptr_tmp = strdup(charptr_tmp); }
+        if (0 != setenv(env_var_name,new_value,1)) {
+            throw error_setting_environment {"cannot change environment variable: "s + env_var_name};
+        }
+    }
+
+    void restore_environment_variable (const char *env_var_name, char* & charptr_tmp) {
+        if (charptr_tmp) {
+            if (0 != setenv(env_var_name, charptr_tmp, 1)) {
+                throw error_setting_environment {"cannot restore environment variable: "s + env_var_name};
+            }
+            free(charptr_tmp);
+            charptr_tmp = nullptr;
+        }
+    }
+}
+
+TEST_CASE("test atomic metadata with BinBytesBuf", "[xml]")
+{
+    char* save_protocol;
+
+    change_environment_variable("irodsProt",save_protocol,"1");
+
+    load_client_api_plugins();
+
+    auto conn_pool = irods::make_connection_pool(2);
+
+    rodsEnv env;
+    _getRodsEnv(env);
+
+    const auto sandbox = fs::path{env.rodsHome} / "unit_testing_sandbox";
+    auto conn = conn_pool->get_connection();
+
+    REQUIRE(fs::client::create_collection(conn, sandbox));
+    const auto json_input = json{
+        {"entity_name", sandbox.string()},
+        {"entity_type", "collection"},
+        {"operations", json::array({
+            {
+                {"operation", "add"},
+                {"attribute", "the_attr"},
+                {"value", "the_val"},
+                {"units", "the_units"}
+            },
+            {
+                {"operation", "add"},
+                {"attribute", "name"},
+                {"value", "john <&doe&>"},
+                {"units", ""}
+            }
+        })}
+    }.dump();
+
+    char* json_error_string{};
+    irods::at_scope_exit free_memory{[&json_error_string] { std::free(json_error_string); }};
+
+    auto* conn_ptr = &static_cast<rcComm_t&>(conn);
+    REQUIRE(rc_atomic_apply_metadata_operations(conn_ptr, json_input.c_str(), &json_error_string) == 0);
+    REQUIRE(json_error_string == "{}"s);
+
+    restore_environment_variable("irodsProt",save_protocol);
+
+    const auto gql = fmt::format("select COUNT(COLL_ID) where COLL_NAME = '{}' and META_COLL_ATTR_VALUE like '{}'",
+                                 sandbox.string().c_str(), "%" "<&" "%" "&>" "%");
```

Comment:
Why chop the third argument to `fmt::format` like this? Just curious.

Reply:
I thought it had better clarity, but it's a subjective thing of course.

Reply:
left strings in concatenated form but parenthesized to make clear it's a single argument